contracts/src/Summa.sol

```diff
@@ -77,27 +57,89 @@ contract Summa is Ownable {
     );
     event LiabilitiesCommitmentSubmitted(
         uint256 indexed timestamp,
-        bytes proof,
-        Cryptocurrency[] cryptocurrencies
+        bytes proof
     );
 
     constructor(
         address _verificationKey,
         IVerifier _polynomialEncodingVerifier,
-        uint16 currenciesCount,
+        string[] memory cryptocurrencyNames,
+        string[] memory cryptocurrencyChains,
         uint8 balanceByteRange
     ) {
         require(
             _verificationKey != address(0),
             "Invalid verifying key address"
         );
         verificationKey = _verificationKey;
+        require(
+            cryptocurrencyNames.length == cryptocurrencyChains.length,
+            "Cryptocurrency names and chains number mismatch"
+        );
+        for (uint i = 0; i < cryptocurrencyNames.length; i++) {
+            require(
+                bytes(cryptocurrencyNames[i]).length != 0 &&
+                    bytes(cryptocurrencyChains[i]).length != 0,
+                "Invalid cryptocurrency"
+            );
+        }
+        require(
+            validateVKPermutationsLength(
+                _verificationKey,
+                cryptocurrencyNames.length
+            ),
+            "The number of cryptocurrencies does not correspond to the verifying key"
+        );
         require(
             address(_polynomialEncodingVerifier) != address(0),
             "Invalid polynomial encoding verifier address"
         );
         polynomialEncodingVerifier = _polynomialEncodingVerifier;
-        config = SummaConfig(currenciesCount, balanceByteRange);
+        require(balanceByteRange > 0, "Invalid balance byte range");
+        config = SummaConfig(
+            cryptocurrencyNames,
+            cryptocurrencyChains,
+            balanceByteRange
+        );
+    }
+
+    /**
+     * @dev Validate the number of permutations in the verifying key
+     * @param vkContract The address of the verifying key contract
+     * @param numberOfCurrencies The number of cryptocurrencies whose polynomials are committed in the proof
+     * @return isValid True if the number of permutations in the verifying key corresponds to the number of cryptocurrencies
+     */
+    function validateVKPermutationsLength(
+        address vkContract,
+        uint256 numberOfCurrencies
+    ) internal view returns (bool isValid) {
+        // The number of permutations is 2 + 4 * numberOfCurrencies because of the circuit structure:
+        // 1 per instance column, 1 per constant column (range check) and 4 per range check columns times the number of currencies
+        uint256 numPermutations = 2 + 4 * numberOfCurrencies;
+
+        uint256 startOffsetForPermutations = 0x2e0; // The value can be observed in the VerificationKey contract, the offset is pointing after all the parameters and the fixed column commitment
```

Comment:
For record, this position may changed at recent `halo2-solidity-verifier`
related issue: https://github.com/summa-dev/halo2-solidity-verifier/issues/1